hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/commit/staging/StagingCommitter.java

```diff
@@ -0,0 +1,908 @@
+      // which returns true if there is >1 entry under a given path.
+      FileStatus[] stats = fs.listStatus(attemptPath);
+      LOG.debug("{} files to commit under {}", stats.length, attemptPath);
+      return stats.length > 0;
+    } catch (FileNotFoundException e) {
+      // list didn't find a directory, so nothing to commit
+      // TODO: throw this up as an error?
+      LOG.info("No files to commit");
+      throw e;
+    }
+  }
+
+  @Override
+  public void commitTask(TaskAttemptContext context) throws IOException {
+    try (DurationInfo d = new DurationInfo(LOG,
+        "%s: commit task %s", getRole(), context.getTaskAttemptID())) {
+      int count = commitTaskInternal(context, getTaskOutput(context));
+      LOG.info("{}: upload file count: {}", getRole(), count);
+    } catch (IOException e) {
+      LOG.error("{}: commit of task {} failed",
+          getRole(), context.getTaskAttemptID(), e);
+      getCommitOperations().taskCompleted(false);
+      throw e;
+    }
+    getCommitOperations().taskCompleted(true);
+  }
+
+  /**
+   * Commit the task by uploading all created files and then
+   * writing a pending entry for them.
+   * @param context task context
+   * @param taskOutput list of files from the output
+   * @return number of uploads committed.
+   * @throws IOException IO Failures.
+   */
+  protected int commitTaskInternal(final TaskAttemptContext context,
+      List<FileStatus> taskOutput)
+      throws IOException {
+    LOG.debug("{}: commitTaskInternal", getRole());
+    Configuration conf = context.getConfiguration();
+
+    final Path attemptPath = getTaskAttemptPath(context);
+    FileSystem attemptFS = getTaskAttemptFilesystem(context);
+    LOG.debug("{}: attempt path is {}", getRole(), attemptPath);
+
+    // add the commits file to the wrapped committer's task attempt location.
+    // of this method.
+    Path commitsAttemptPath = wrappedCommitter.getTaskAttemptPath(context);
+    FileSystem commitsFS = commitsAttemptPath.getFileSystem(conf);
+
+    // keep track of unfinished commits in case one fails. if something fails,
+    // we will try to abort the ones that had already succeeded.
+    int commitCount = taskOutput.size();
+    final Queue<SinglePendingCommit> commits = new ConcurrentLinkedQueue<>();
+    LOG.info("{}: uploading from staging directory to S3", getRole());
+    LOG.info("{}: Saving pending data information to {}",
+        getRole(), commitsAttemptPath);
+    if (taskOutput.isEmpty()) {
+      // there is nothing to write. needsTaskCommit() should have caught
+      // this, so warn that there is some kind of problem in the protocol.
+      LOG.warn("{}: No files to commit", getRole());
+    } else {
+      boolean threw = true;
+      // before the uploads, report some progress
+      context.progress();
+
+      PendingSet pendingCommits = new PendingSet(commitCount);
+      try {
+        Tasks.foreach(taskOutput)
```

Comment:
We added retry to `Tasks` recently, which helps reliability. Here's our implementation, which is based on the exponential back-off code from Presto:

```java
    private <E extends Exception> void runTaskWithRetry(Task<I, E> task, I item)
        throws E {
      long start = System.currentTimeMillis();
      int attempt = 0;
      while (true) {
        attempt += 1;
        try {
          task.run(item);
          break;

        } catch (Exception e) {
          LOG.warn("Retrying task after failure: " + e.getMessage());

          long durationMs = System.currentTimeMillis() - start;
          if (attempt >= maxAttempts || durationMs > maxDurationMs) {
            throw e;
          }

          for (Class<? extends Exception> exClass : stopRetryExceptions) {
            if (exClass.isInstance(e)) {
              throw e;
            }
          }

          int delayMs = (int) Math.min(
              minSleepTimeMs * Math.pow(scaleFactor, attempt - 1),
              maxSleepTimeMs);
          int jitter = ThreadLocalRandom.current()
              .nextInt(Math.max(1, (int) (delayMs * 0.1)));

          try {
            TimeUnit.MILLISECONDS.sleep(delayMs + jitter);
          } catch (InterruptedException ie) {
            Thread.currentThread().interrupt();
            throw new RuntimeException(ie);
          }
        }
      }
    }
```

Reply:
I am going to steer clear of it for now, as there is retry logic deeper in the commit options and FS itself which tries to disguishable between retryable and unrecoverable failures (ConnectionRefused vs UnknownHost, ...); it's happening further down. But I could life this code (and whatever setsup the stopRetryException list if you think it should all go together

Reply:
Up to you. This is more important to us since we use the S3 client directly. If you're using a mechanism in S3A, then you can probably add retries elsewhere. I would recommend having them, though. We see throttling from S3, which shouldn't take down a job.

The `stopRetryException` list is per execution and is seeded with a `NotRetryableException` that can be thrown in the body of run. That way you can identify certain exceptions (like 404 errors in S3) that won't be retried.

Reply:
I'll leave our retrying in the s3a code for now; there's also whatever goes on in the AWS SDK, which we aren't fully understanding of. Ideally I'd like to hook up the S3a instrumentation & stats to get numbers of what things are being retried in the transfer manager. Work for another day